sdk-core-protos/protos/local/temporal/sdk/core/workflow_activation/workflow_activation.proto

```diff
@@ -19,6 +19,32 @@ import "temporal/sdk/core/common/common.proto";
 
 // An instruction to the lang sdk to run some workflow code, whether for the first time or from
 // a cached state.
+//
+// ## Job ordering guarantees and semantics
+//
+// Core will, by default, order jobs within the activation as follows:
+// `patches -> signals/updates -> other -> queries -> evictions`
+//
+// This is because:
+// * Patches are expected to apply to the entire activation
+// * Signal and query handlers should be invoked before workflow routines are iterated
+// * Queries always go last (and, in fact, always come in their own activation)
+//
+// The downside of this reordering is that a signal or update handler may not observe that some
+// other event had already happened (ex: an activity completed) when it is first invoked, though it
+// will subsequently when workflow routines are driven. Core only does this reordering to make life
```

Comment:
>  a signal or update handler may not observe that some other event had already happened (ex: an activity completed) when it is first invoked

OK, this makes sense: the activity completion is in the "other" category.

> though it will subsequently when workflow routines are driven.

Struggling with this. Does "driving workflow routines" happen when jobs in the "other" category are handled? If so, the update handler has stopped executing by that point hasn't it, so in what sense is it able to "observe" something that it didn't when it first executed?

Reply:
Yes, it does (though, per what I said in that other issue, ideally it happens at the end of processing all the jobs rather than inline with it).

The handler counts as one of the workflow routines, and so is again allowed to advance at the end of job processing. At that point, it would see the activity as completed, if it hadn't already exited.